src/v/redpanda/admin/services/cluster.cc

```diff
@@ -51,11 +81,29 @@ cluster_service_impl::list_kafka_connections(
 
     auto units = co_await kcs.rate_limit();
 
+    auto max_connections = calculate_max_connections();
+    if (max_connections == 0) {
+        throw_no_memory(max_connections);
+    }
+
+    // Adjust page size according to available memory, but still try to return a
+    // response to support page_size=INT_MAX requests with a low connection
+    // count.
+    auto original_page_size = kcs.get_effective_limit(req.get_page_size());
+    auto new_page_size = std::min(original_page_size, max_connections);
+    req.set_page_size(static_cast<int32_t>(new_page_size));
+
     auto resp = ctx.is_proxied()
                   ? co_await kcs.list_kafka_connections_local(std::move(req))
                   : co_await kcs.list_kafka_connections_cluster_wide(
                       _proxy_client, ctx, std::move(req));
 
+    if (
+      new_page_size < original_page_size
+      && resp.get_total_size() > new_page_size) {
+        throw_no_memory(max_connections);
```

Comment:
I'm not sure I understand the reasoning here.

Reply:
OK. I think it would be useful to return a subset, but if that's how to API is designed, then that's fine by me.